consensus/istanbul/backend/proxy_handler.go

```diff
@@ -0,0 +1,692 @@
+			j := 0
+			for val := range validators {
+				valAddresses[j] = val
+				j++
+			}
+			proxies[i].Validators = valAddresses
+		}
+
+		i++
+	}
+	return proxies
+}
+
+// assignmentPolicy is intended to allow different
+// solutions for assigning validators to proxies
+type assignmentPolicy interface {
+	addProxy(proxy *proxy)
+	removeProxy(proxy *proxy)
+
+	getValAssignments() *valAssignments
+
+	addValidators(map[common.Address]bool)
+	removeValidators(map[common.Address]bool)
+
+	reassignValidators()
+}
+
+type hasher struct{}
+
+func (h hasher) Sum64(data []byte) uint64 {
+	return xxhash.Sum64(data)
+}
+
+// consistentHashingPolicy uses consistent hashing to assign validators to proxies.
+// Validator <-> proxy pairings are recalculated every time a proxy or validator
+// is added/removed
+type consistentHashingPolicy struct {
+	c              *consistent.Consistent // used for consistent hashing
+	valAssignments *valAssignments
+}
+
+func newConsistentHashingPolicy() *consistentHashingPolicy {
+	// This sets up a consistent hasher with bounded loads:
+	// https://ai.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html
+	// Partitions are assigned to members (proxies in this case)
+	// using a hash ring.
+	// When locating a key's member using `LocateKey`, the key is assigned
+	// to a partition using hash(key) % PartitionCount in constant time.
+	cfg := consistent.Config{
+		// Prime to distribute keys more uniformly.
+		// Higher partition count generally gives a more even distribution
+		PartitionCount:    271,
+		// The number of replications of a member (proxy) on the hash ring
+		ReplicationFactor: 40,
+		// Used to enforce a max # of partitions assigned per member, which is
+		// (PartitionCount / len(members)) * Load. A load closer to 1 gives
+		// more uniformity in the # of partitions assigned to specific members,
+		// but a higher load results in less relocations when members are added/removed
+		Load:              1.2,
+		Hasher:            hasher{},
+	}
+
+	return &consistentHashingPolicy{
+		valAssignments: newValAssignments(),
+		c: consistent.New(nil, cfg),
+	}
+}
+
+// getValAssignments returns the current validator assignments
+func (ch *consistentHashingPolicy) getValAssignments() *valAssignments {
+	return ch.valAssignments
+}
+
+// addProxy adds a proxy to the consistent hasher and recalculates all validator assignments
+func (ch *consistentHashingPolicy) addProxy(proxy *proxy) {
+	ch.c.Add(proxy.ID())
+	ch.reassignValidators()
+}
+
+// removeProxy removes a proxy from the consistent hasher and recalculates all validator assignments
+func (ch *consistentHashingPolicy) removeProxy(proxy *proxy) {
+	ch.c.Remove(proxy.ID().String())
+	ch.reassignValidators()
+}
+
+// addValidators adds validators to the valAssignments struct and recalculates
+// all validator assignments
+func (ch *consistentHashingPolicy) addValidators(vals map[common.Address]bool) {
+	ch.valAssignments.addValidators(vals)
```

Comment:
@asaj to follow up on your question yesterday about situations when a validator will be in the valAssignments `valToProxy` with a proxy value of nil, it could be if validators are added but no proxies have been added, or during the short period before reassignments are calculated after adding validators